server/src/main/java/org/jetbrains/bsp/bazel/server/sync/BazelProjectMapper.kt

```diff
@@ -0,0 +1,127 @@
+package org.jetbrains.bsp.bazel.server.sync
+
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.runBlocking
+import org.jetbrains.bsp.bazel.info.BspTargetInfo.*
+import org.jetbrains.bsp.bazel.logger.BspClientLogger
+import org.jetbrains.bsp.bazel.server.bsp.utils.SourceRootGuesser
+import org.jetbrains.bsp.bazel.server.sync.dependencytree.DependencyTree
+import org.jetbrains.bsp.bazel.server.sync.languages.LanguagePluginsService
+import org.jetbrains.bsp.bazel.server.sync.model.*
+import org.jetbrains.bsp.bazel.workspacecontext.WorkspaceContext
+import java.net.URI
+
+class BazelProjectMapper(
+    private val languagePluginsService: LanguagePluginsService,
+    private val bazelPathsResolver: BazelPathsResolver,
+    private val targetKindResolver: TargetKindResolver,
+    private val logger: BspClientLogger,
+) {
+    fun createProject(
+        targets: Map<String, TargetInfo>,
+        rootTargets: Set<String>,
+        workspaceContext: WorkspaceContext
+    ): Project {
+        languagePluginsService.prepareSync(targets.values.asSequence())
+        val dependencyTree = DependencyTree(rootTargets, targets)
+        val targetsToImport = selectTargetsToImport(workspaceContext, rootTargets, dependencyTree)
+        val modulesFromBazel = createModules(targetsToImport, dependencyTree)
+        val workspaceRoot = bazelPathsResolver.workspaceRoot()
+        val syntheticModules =
+            createSyntheticModules(modulesFromBazel, workspaceRoot, workspaceContext)
+        val allModules = modulesFromBazel + syntheticModules
+        val sourceToTarget = buildReverseSourceMapping(modulesFromBazel)
+        return Project(workspaceRoot, allModules.toList(), sourceToTarget)
+    }
+
+    private fun selectTargetsToImport(
+        workspaceContext: WorkspaceContext, rootTargets: Set<String>, tree: DependencyTree
+    ): Sequence<TargetInfo> = tree.allTargetsAtDepth(
+        workspaceContext.importDepth.value, rootTargets
+    ).asSequence().filter(::isWorkspaceTarget)
+
+
+    private fun isWorkspaceTarget(target: TargetInfo): Boolean = target.id.startsWith("//")
+
+    private fun createModules(
+        targetsToImport: Sequence<TargetInfo>, dependencyTree: DependencyTree
+    ): Sequence<Module> = runBlocking {
+        targetsToImport.toList().map { target: TargetInfo ->
+            async {
+                createModule(target, dependencyTree).takeUnless {
+                    it.tags.contains(Tag.NO_IDE)
+                }
+            }
+        }.awaitAll().asSequence().filterNotNull()
```

Comment:
i think that we can use flow here - it'd be something like that:
```suggestion
targetsToImport.asFlow()
            .map { createModule(it, dependencyTree) }
            .filterNot { it.tags.contains(Tag.NO_IDE) }
            .toList()
            .asSequence()
```